examples/proxy/lib.rs

```diff
@@ -0,0 +1,120 @@
+//! This example demonstrates how the Proxy/Forward pattern can be
+//! implemented in ink!.
+//!
+//! What the contract does is:
+//!
+//!   * Any call to this contract that does not match a selector
+//!     of it is forwarded to a specified address.
+//!   * The instantiator of the contract can modify this specified
+//!     `forward_to` address at any point.
+//!
+//! Using this pattern it is possible to implement upgradable contracts.
+//!
+//! Note though that the contract to which calls are forwarded still
+//! contains it's own state.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use ink_lang as ink;
+
+#[ink::contract]
+pub mod proxy {
+    /// A simple proxy contract.
+    #[ink(storage)]
+    pub struct Proxy {
+        /// The `AccountId` of a contract where any call that does not match a
+        /// selector of this contract is forwarded to.
+        forward_to: AccountId,
+        /// The `AccountId` of a privileged account that can update the
+        /// forwarding address. This address is set to the account that
+        /// instantiated this contract.
+        admin: AccountId,
+    }
+
+    /// Error which is returned if a caller tries to invoke
+    /// [`Proxy::change_forward_address`] while not being the
+    /// admin account stored in the storage of this contract.
+    #[derive(scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub struct NotAdminErr {
+        admin: AccountId,
+    }
+
+    impl core::fmt::Display for NotAdminErr {
+        fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+            let msg = ink_env::format!(
+                "caller does not have sufficient permissions, only {:?} does",
+                self.admin
+            );
+            write!(f, "{}", msg)
+        }
+    }
+
+    impl Proxy {
+        /// Instantiate this contract with an address of the `logic` contract.
+        ///
+        /// Sets the privileged account to the caller. Only this account may
+        /// later changed the `forward_to` address.
+        #[ink(constructor)]
+        pub fn new(forward_to: AccountId) -> Self {
+            Self {
+                admin: Self::env().caller(),
+                forward_to,
+            }
+        }
+
+        /// Changes the `AccountId` of the contract where any call that does
+        /// not match a selector of this contract is forwarded to.
+        ///
+        /// Returns a [`NotAdminErr`] in case the caller is not the
+        /// admin account stored in the storage of this contract.
+        ///
+        /// # Note:
+        ///
+        /// Returning an `Result::Err` here will result in the transaction
+        /// being reverted, in that sense it is equal to e.g. placing an
+        /// `assert!` here instead.
+        #[ink(message)]
+        pub fn change_forward_address(
+            &mut self,
+            new_address: AccountId,
+        ) -> Result<(), NotAdminErr> {
```

Comment:
@Robbepop While running the waterfall tests for this example, I found that since I switched from `assert!` to returning `Result::Err` the transaction is no longer reverted ‒ it always results in `ExtrinsicFailed`.

If you don't see anything that I missed I'll consider this a bug, revert to `assert!` and create a follow-up issue.

Reply:
Hmm, I tested that on Europa 0.3.6 and it is reverted.
<img width="1482" alt="image" src="https://user-images.githubusercontent.com/18346821/143215447-0518b619-a89d-48bb-8b8c-87d84524e2a3.png">

Another question, how the user can forward some data via polkadot-js?=) Because I can only read that=D
<img width="1285" alt="image" src="https://user-images.githubusercontent.com/18346821/143215795-453957ff-6b10-4f4d-896e-fe4a2aaa2a5c.png">

Reply:
Hmm, could be that `substrate-contracts-node` is somewhat outdated.

Reply:
Found it, the UI doesn't reflect this yet: https://github.com/polkadot-js/apps/issues/6465.

>Another question, how the user can forward some data via polkadot-js?=) Because I can only read that=D

@xgreenx You have to switch the metadata of this contract to some other contract. This can be done by clicking the "Add an existing contract" button, entering the proxy smart contract address and upload a different `.contract` file there. The UI will then use the ABI of this contract to interact with the proxy.